src/domain/accounts/services/src/authentication_service_impl.rs

```diff
@@ -136,16 +129,50 @@ impl AuthenticationServiceImpl {
     pub fn decode_access_token(
         &self,
         access_token: &str,
-    ) -> Result<TokenData<KamuAccessTokenClaims>, AccessTokenError> {
+    ) -> Result<AccessTokenArg, AccessTokenError> {
+        if &access_token[..2] == ACCESS_TOKEN_PREFIX {
+            return self
+                .access_token_svc
+                .decode_access_token(access_token)
+                .map_err(|err| AccessTokenError::Invalid(Box::new(err)))
+                .map(AccessTokenArg::KamuAccessToken);
+        }
         let mut validation = Validation::new(KAMU_JWT_ALGORITHM);
         validation.set_issuer(&[KAMU_JWT_ISSUER]);
 
-        decode::<KamuAccessTokenClaims>(access_token, &self.decoding_key, &validation).map_err(
-            |e| match *e.kind() {
+        decode::<KamuAccessTokenClaims>(access_token, &self.decoding_key, &validation)
+            .map_err(|e| match *e.kind() {
                 ErrorKind::ExpiredSignature => AccessTokenError::Expired,
                 _ => AccessTokenError::Invalid(Box::new(e)),
-            },
-        )
+            })
+            .map(AccessTokenArg::JWTToken)
+    }
+
+    pub async fn account_by_token_impl(
+        &self,
+        access_token: &str,
+    ) -> Result<Account, GetAccountInfoError> {
+        let decoded_access_token = self
+            .decode_access_token(access_token)
+            .map_err(GetAccountInfoError::AccessToken)?;
+
+        match decoded_access_token {
+            AccessTokenArg::JWTToken(token_data) => {
+                let account_id = AccountID::from_did_str(&token_data.claims.sub)
+                    .map_err(|e| GetAccountInfoError::Internal(e.int_err()))?;
+
+                match self.account_by_id(&account_id).await {
+                    Ok(Some(account)) => Ok(account),
+                    Ok(None) => Err(GetAccountInfoError::AccountUnresolved),
+                    Err(e) => Err(GetAccountInfoError::Internal(e)),
+                }
+            }
+            AccessTokenArg::KamuAccessToken(kamu_access_token) => self
+                .access_token_svc
+                .find_account_id_by_active_token_id(&kamu_access_token.id)
```

Comment:
Exactly what I meant. We only require an ID, no need to reconstruct the full token structure.